Audio Extraction Service + extra features

greenw0lf opened the pull request:
Linked to #27 

This PR adds API code to run the audio extraction worker as a service (on the cluster). Also added some extra features that also exist in the Whisper ASR worker.

Notable changes:
- Added the same scripts as in the Whisper ASR worker (that Govert made)
- Small cleanup and reformatting of the Dockerfile
- API code (in `api.py`) that functions similar to the Whisper ASR worker
- Provenance logging and uploading to S3
- Removing all input/intermediary/output files if the worker fails/succeeds
- Saving both input & output in one folder (with its name being the `carrier_id` of the input)

I have done some manual tests of 3 videos from OpenBeelden and a random video from the internet so far and it all seems to run smoothly, so I think this is ready for review

@Veldhoen or @jblom , could you have a look at this fully?
@gb-beng @mwigham @KleinRana @wmelder maybe of interest to you (tagging you since we discussed in today's gathering to do this more often)

greenw0lf committed:
Initial "AE worker as a service" work

greenw0lf committed:
Address formatting and mypy issues

greenw0lf committed:
Ready to test on the cluster now

greenw0lf committed:
Comment ports in docker compose

greenw0lf committed:
Remove service flag from entrypoint

greenw0lf committed:
Update workflows to same versions as whisper ASR worker

greenw0lf committed:
Change in workflow to trigger deployment

greenw0lf committed:
Remove all files if pipeline succeeds

Accidentally not added, now it should work.

greenw0lf commented:
Also, forgot to mention

You can do some manual testing yourself if you want/need to at https://audio-extract.apps.ostest.beeldengeluid.nl/docs

Veldhoen commented on api.py:
```diff
@@ -0,0 +1,194 @@
+        task.status = Status.ERROR if error_msg else Status.DONE
+        task.error_msg = error_msg
+    except Exception:
+        logger.exception("Failed to run audio extraction")
```
Maybe add error message from Exception to the logger

Veldhoen commented on download.py:
```diff
@@ -17,55 +20,102 @@
-    input_file = os.path.join(input_file_dir, fn)
-    # TODO rethink how to return mime_type
-    _, extension = get_asset_info(input_file)
+    asset_id, extension = get_asset_info(fn)
```
I think the worker could (and should) be agnostic to the concept of asset_ids: it only knows that it's dealing with an input file (S3 object) and an output file (S3 object). 
The asset_id is only used for choosing file paths, but you might as well use the task id for that (or, if it seems relevant, the (prefix of) the filename)
The extension is only needed inside the ffmpeg_audio_extraction, so you can obtain it there.

greenw0lf replied:
I disagree because the `task ID` should be reserved only for the API part. I want some sort of separation between the API part and the worker's actual processing part if that makes sense. Also, it would add more input for the worker part to process, whereas the `asset_id` is something that can simply be obtained from the input file's name.

I could indeed not use the asset_id and just use generic names like `input.mp4`, but if processing failed and the files were not deleted somehow, they would persist there and create issues with processing the next tasks, leading to a total worker failure. This can happen, for example, when trying to download a file but not having full access to the URL domain (like was the case when testing with input files from `modelhosting.beng.nl`).

Veldhoen replied:
Ahh yes, very fair point about the task ID being just for the API part. I think what you are doing in practice is using the basename of the file, which makes sense. I just wouldn't call it `asset_id`, because a filename does not necesarily correspond to an id.

Veldhoen commented on download.py:
```diff
@@ -17,55 +20,102 @@
-    if not validate_s3_uri(s3_uri):
-        logger.error(f"Invalid S3 URI: {s3_uri}")
-        return None
+    provenance = {
```
Maybe specify a (data)class for the provenance, so you only need to fill in the relevant fields here, and can pass it around: that way, you don't have to obtain any information from the config in this module

Veldhoen replied:
Also I think provenance steps may be a bit irrelevant/overkill here - although it is nice to keep track of both worker code version and ffmpeg version

greenw0lf replied:
I agree with the dataclass idea, but not so much with the provenance steps being overkill. Since we use that for the Whisper worker, I think it would make sense to also use it for this worker, have something that is standard between all workers. Plus, I would only change this once we have a proper discussion within the team about what the format of the provenance should be. For now, we agreed to keep using this older format which seems to cover our needs, so no need to oversimplify here.

Veldhoen commented on api.py:
```diff
@@ -0,0 +1,194 @@
+
+class Task(BaseModel):
+    input_uri: str
+    output_uri: str
```
From the code, I think the output_uri should in fact be an S3 prefix rather than an entire uri, where both the output audio and provenance will be uploaded to. I think it would be nice to make the worker really agnostic, and have the caller of the API explicitly specify where the provenance should be stored.

greenw0lf replied:
I don't get this feedback. Can you explain/reformulate?

Veldhoen replied:
Technically, `output_uri` is not a uri: it doesn't point to a single object. 
Instead, `output_uri` is the destination _folder_ in S3 where both the output audio and provenance will be uploaded to.
And technically in S3, a 'folder' does not exist: it is just a _prefix_ that contains `/` characters.

Veldhoen replied:
And the provenance filename (as well as the audio filename, for that matter) are configured, the caller (from Airflow, in this case) can only guess the actual uris of the output objects (from the prefix). It would be nice to make them explicit.

Veldhoen commented on main.py:
```diff
@@ -2,7 +2,6 @@
 import sys
 from base_util import LOG_FORMAT
 from config import input_uri, output_uri
```
Why not leave these out of the configuration?

greenw0lf replied:
This was from prior code when we were thinking of running this worker as a job (which sadly isn't possible). I could remove it and have the worker be exclusively a service at this point since the job isn't used for anything other than testing (which even could be tested via the service way).

Veldhoen commented on run_pipeline.py:
```diff
@@ -1,63 +1,107 @@
     result = download_uri(input_uri)
     logger.info(result)
-    if not result:
+    if result.error != "":
```
Cleaner to have a try/except clause (have the function raise an exception if download fails)

Veldhoen commented on api.py:
```diff
@@ -0,0 +1,194 @@
+    error_msg: str | None = None
+
+
+all_tasks = [
```
Why not keep track of tasks in a dict (per id)? That would save the get_task_by_id and get_task_index etc.

Veldhoen commented on base_util.py:
```diff
@@ -51,3 +54,49 @@ def run_shell_command(command: List[str]) -> bool:
+    try:
+        # write ae_provenance.json
+        with open(
+            os.path.join(output_dir, PROVENANCE_JSON_FILE), "w+", encoding="utf-8"
```
Why is the PROVENANCE_JSON_FILE a global variable? Might as well have it an argument to the function, right? (with a default set)

Veldhoen commented on download.py:
```diff
@@ -17,55 +20,102 @@
-    _, extension = get_asset_info(input_file)
+    asset_id, extension = get_asset_info(fn)
+
+    input_file_dir = os.path.join(data_base_dir, asset_id)
```
You define this here, but you want it to be the same as the output_dir defined elsewhere. So it would be better to create directory and pass it to the download_uri function as an argument

greenw0lf replied:
I prefer having the `download_uri` function do all the download and input-related logic and keep the `run` function in `run_pipeline` as simple as possible (only focus on getting the right input+output from each step and passing it along)

Veldhoen commented on download.py:
```diff
@@ -17,55 +20,102 @@
             os.makedirs(input_file_dir)
         with open(input_file, "wb") as file:
             response = requests.get(url)
+            if response.status_code >= 400:
```
If it's not 200, right?

Veldhoen commented on run_pipeline.py:
```diff
@@ -1,63 +1,107 @@
+from config import (
+    data_base_dir,
+    s3_endpoint_url,
+    ae_samplerate_hz,
```
Not a must, but we could consider passing things like sample rate as body to the APIrequest, rather than obtain them from the config. The service would be more flexible that way.

Veldhoen commented on api.py:
```diff
@@ -0,0 +1,194 @@
+        logger.exception("Failed to run audio extraction")
+        task.status = Status.ERROR
+    update_task(task)
+    logger.info(f"Done extracting audio for task {task.id}")
```
Maybe move this into the try clause (if it fails, it won't really be 'Done')

Veldhoen commented on run_pipeline.py:
```diff
@@ -1,63 +1,107 @@
 
-def run(input_uri: str, output_uri: str) -> bool:
+def run(input_uri: str, output_uri: str) -> Optional[str]:
     logger.info(f"Processing {input_uri} (save to --> {output_uri})")
```
Perhaps do an early check to see whether the output_uri exists already (for instance, if the staged asset is already an mp3 file). Might need to check that the object corresponding to the output_uri fulfills some criteria (sample rate, etc.)

greenw0lf replied:
1. `output_uri` is a path to a directory, not a file. The string of `output_uri` will be concatenated with the resulting audio file from the worker, and then that entire URI will be used to upload the output to S3.
2. Let's say there is an audio file already in that directory. How will I look for it AND also find out info about sample rate, etc? I think this would add additional unnecessary logic to the worker. We assume that there is no audio present in `output_uri` and, if there is, it will be overwritten. The only check that makes sense is for the input to not be audio already. Maybe I can also go beyond that and try to download it, then extract the `sample_rate` and whether it's mono audio or not, then apply ffmpeg if those parameters don't match as well.

Veldhoen replied:
Like we discussed in the office: 
Airflow should be responsible for checking whether the audio file exists already, the worker should not need to bother.
I'm not sure it's relevant to check sample rate/ channels at this point, if I understood correctly Whisper will make sure of this. So I guess we can leave this for now.

Veldhoen reviewed (changes requested):
Generally looks good, I left some comments to improve the readability/maintainability of the code mostly. Most strikingly (I know I mentioned this earlier, sorry), but I really think it's better to use Python's built-in exception handling than do string comparisons or set and check error states. Also, some of the functions are a bit complex/ have to many responsibilities, so could do with a bit of refactoring. Let me know if you need more guidance to tackle this!

greenw0lf committed:
Work for the past 2 days

Lots of refactoring; will summarize in the PR

greenw0lf committed:
Changes for project check pass

greenw0lf committed:
Hopefully finally done with this refactor

greenw0lf committed:
Revert docker-compose change

greenw0lf commented:
I have done quite the refactor in the code, so I'll try to summarize that as well as I can to facilitate the review process:

- `all_tasks` is no longer a list, but a `dict` that has as key the `task_id` and as object the entire `Task`
- Accessing of the tasks is now done by using `task_id` as a key, instead of searching for the correct `Task` through a list
- Better logging message for when a task has been processed and has failed/succeeded ("Task {task_id} has been updated" instead of "Done with task {task_id}"
- `Provenance` is now a `dataclass` (in `base_util.py`)
- `ffmpeg` version now also logged in the provenance, under the `Audio Extraction` step of the final prov
- Provenance filename is now a `.env` variable (`prov_filename` in code or `PROVENANCE_FILENAME` in `.env`)
- Sample rate and converting to mono removed for now as Whisper handles those automatically (so the worker only converts from a video file to audio file, without any additional processing)
- Moved some input/output directory and filenames/extensions code around (hopefully better now!)
- New system for logging errors (`Exception`s raised left and right :D)
- Worker as a job completely removed, only support for the worker as a service
- Output of a step, if it already exists, will be deleted and overwritten by the step's process
- Output filenames returned as a `response` parameter in the `Task` object (which can then be appended to the `output_uri` to get the full path to each output file)
- `output_uri` can also be empty (for local testing, to keep the input+output files stored instead of deleting them)

Veldhoen commented:
I responded to some older conversations, may not be relevant anymore though :')
Will now have a look at the code!

Veldhoen commented on api.py:
```diff
@@ -0,0 +1,181 @@
+    return True
+
+
+def update_task(task: Task) -> bool:
```
Why return a bool? Might as well raise an exception here

Veldhoen commented on api.py:
```diff
@@ -0,0 +1,181 @@
+current_task: Optional[Task] = None
+
+
+def delete_task(task_id) -> bool:
```
Why return a bool? Might as well raise an exception here

Veldhoen commented on api.py:
```diff
@@ -0,0 +1,181 @@
+    except Exception as e:
+        logger.error("Failed to run audio extraction")
+        logger.exception(e)
+        task.status = Status.ERROR
```
Set error message for task?

Veldhoen commented on base_util.py:
```diff
@@ -28,26 +46,106 @@ def extension_to_mime_type(extension: str) -> str:
 
 # used by transcode.py
-def run_shell_command(command: List[str]) -> bool:
+def run_shell_command(command: List[str], ret_output=False) -> bool:
```
`ret_output` is not used

Veldhoen commented on base_util.py:
```diff
@@ -28,26 +46,106 @@ def extension_to_mime_type(extension: str) -> str:
+    return process.returncode == 0
+
+
+def run_shell_command_with_output(command: List[str]) -> str:
```
Why have two almost identical functions for running shell commands?

greenw0lf replied:
Because I want one to just inform me whether it succeeded or not (thus returning a bool), and I need the output from the other (thus returning a str that I use for the ffmpeg version provenance).

I could return both with one function? So maybe the output can be a list or a tuple where the first element is a bool and the second is a string corresponding to the error/output, or vice versa for the order of the output.

Veldhoen commented on base_util.py:
```diff
@@ -28,26 +46,106 @@ def extension_to_mime_type(extension: str) -> str:
+    output_path: str,
+    output_uri: str,
+    asset_id: str,
+) -> bool:
```
Why return a bool? No return needed; just do it or fail and raise an error

Veldhoen commented on config.py:
Does it make sense to use all those defaults for the config, shouldn't the application just fail if some of those values are missing in the environment?

Veldhoen replied:
BTW, I think we mostly use all-caps for configured values to make them stand out (although I'm not sure we adhere to PEP conventions in that way). See for instance https://github.com/beeldengeluid/frozen-set-api/blob/7-reconcile-queries/src/config.py

greenw0lf replied:
I guess it doesn't make much sense, indeed. I will remove some of those defaults!

And hmm, about the config values, maybe we should discuss this on Slack and see what the convention actually is and apply it to both repos (if you have time to do so, of course)

Veldhoen commented on docker-compose.yml:
```diff
@@ -4,10 +4,12 @@ services:
-    restart: no
\ No newline at end of file
+    restart: no
+    # For testing the service locally
```
Docker compose is for local test only anyway, right?

Veldhoen commented on transcode.py:
```diff
@@ -1,64 +1,80 @@
-def ffmpeg_transcode(input_path, asset_id, extension) -> Optional[str]:
+def ffmpeg_audio_extraction(
+    input_file: str,
+    asset_id: str,
```
I feel asset_id and extension should be obtained within this function, rather than passed as an argument

Veldhoen reviewed (commented):
Lots of improvements, I think the coder is much cleaner this way! Some small comments here and there, nothing breaking though.

greenw0lf committed:
Final set of changes before merging

greenw0lf committed:
uppercase config params

After some discussion, we agreed that config params are considered as PEP 8 constants and the standard for those is to write them all UPPERCASE with _ between words

greenw0lf committed:
Remove deploying image on service branch

Since this will be merged and deployments are meant to run only on main, it doesn't make sense to have them run on other branches. This was here for cluster testing purposes.
